### PR TITLE
[CDRIVER-4637] A config settings command

### PR DESCRIPTION
**Pull request by vector-of-bool**
This changeset is partial work towards CDRIVER-4637. It introduces a CMake module `MongoSettings.cmake` which defines CMake utilities for defining build settings in a reliable, uniform, and expressive manner.

Originally, this changeset was going to also complete CDRIVER-4637 and remove the `AUTO` build settings that are used throughout the project. However, this proved to be an enormous task as many of our CI configurations implicitly relied `AUTO` to make things "just work" regardless of the ambient build environment (this is exactly the thing we are purposefully trying to remove via CDRIVER-4637). For this reason, updating any individual build setting away from `AUTO` is substantial effort to coordinate with the CI executors as to when and where certain features are available.

# Setting Design

The `mongo_setting()` command looks like the following:

```
mongo_setting(
    <name> <doc>
    [TYPE <BOOL|PATH|FILEPATH|STRING>]
    [DEFAULT [[DEVEL|AUDIT] [VALUE <value> | EVAL <code>]] ...]
    [OPTIONS [<opts> ...]]
    [VALIDATE [CODE <code>]]
    [ADVANCED] [ALLOW_OTHER_VALUES]
)
```

The full explanation of all arguments is available in `MongoSettings.cmake`. It basically works like this:

1. `<name>`, `<doc>`, and `TYPE` correspond to the parameters to `set()` in `set(<name> [...] CACHE <TYPE> <doc>)`
2. The `DEFAULT` argument takes one of two forms: `VALUE` and `EVAL`
    1. For `VALUE <value>`, the given `<value>` is set as the default value. Simple.
    2. For `EVAL <code>`, the given `<code>` is executed in order to determine the default value. This is used where the default value should depend on certain conditions about the system, such as the operating system and host architecture. Importantly, it SHOULD NOT depend on things like the result of a `find_package()` or ambient system state. This is designed to replace uses of `AUTO` to toggle operating-system-specific settings. Currently, it is only used in one location, to turn on `ENABLE_SHM_COUNTERS` in the same conditions under which `AUTO` would have enabled this feature (this is the only feature that I have found does not generate trouble when removing `AUTO`).
3. The `OPTIONS` argument specifies a list of possible values that may be taken by the build setting. This also sets the `STRINGS` property on the cache variable. If provided, `mongo_setting` will validate that the build setting is one of these values unless `ALLOW_OTHER_VALUES` is also given.
4. `ADVANCED` will mark the cache variable as advanced.
5. `VALIDATE CODE <code>` will execute `<code>` after the cache value has been set. This code should issue an error/warning if the value is incompatible with another setting or with the toolchain/platform.
6. The `DEVEL` and `AUDIT` toggles to `DEFAULT` allow different default default values to be set when certain environment variables are present. This isn't strictly related to CDRIVER-4637, but is a convenience for developers and CI.

Additionally, there is a `mongo_bool_setting()` command that defines a build setting of type `BOOL` with an implicit default of `TRUE`.

# CMake Version Updates 

This changeset requires a more recent (but still relatively old) CMake version. Several build tasks were still using prior `find-cmake` logic. Those have been updated to use `find-cmake-latest`.

It remains that the build-with-toolchain uses an old CMake version that conflicts with the newer CMake required. Addressing this is an open problem.

# MinGW Changes

The MinGW build tasks invoked a `.bat` script that hardcoded a path to CMake and GCC. The CMake that it used is too old to work with these changes, and batch is undoubtedly the worse scripting language ever conceived and widely adopted.

Since we now have the technology, the MinGW task has been modified to stay within Cygwin Bash and invoke the CMake command correctly to build with MinGW GCC. This removes the hardcoding of the (ancient 4.9) MinGW GCC version, and CMake now finds a more modern GCC 8 available on the PATH. Ideally, we would not rely on an arbitrary GCC being available for this process, so I opened CDRIVER-4652 to address this, hopefully sooner than later.

### CODE REVIEW

File: build/cmake/MongoSettings.cmake [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        <name> <doc>
+        [DEFAULT [[DEVEL|AUDIT] [VALUE <value> | EVAL <code>]] ...]
+        [VALIDATE [CODE <code>]]
+        [ADVANCED] [ALLOW_OTHER_VALUES]
+    )
+
+This is a shorthand for defining a boolean setting. See mongo_setting() for more
+option information. The TYPE of the setting will be BOOL, and the implicit
+default value for the setting will be ON if neither DEFAULT nor DEFAULT_EVAL are
+specified.
```
Comment (kevinAlbs) [inline build/cmake/MongoSettings.cmake:260]:
```suggestion
default value for the setting will be ON if DEFAULT is not specified.
```

Appears to be no `DEFAULT_EVAL` option.

File: build/cmake/MongoSettings.cmake [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+]==]
+function(mongo_bool_setting name doc)
+    set(args ${ARGN})
+    # Inject "ON" as a default:
+    if(NOT "DEFAULT" IN_LIST args AND NOT "DEFAULT_EVAL" IN_LIST args)
```
Comment (kevinAlbs) [inline build/cmake/MongoSettings.cmake:266]:
```suggestion
    if(NOT "DEFAULT" IN_LIST args)
```

File: src/libmongoc/CMakeLists.txt [hunk 3]
```diff
@@ -372,25 +372,7 @@ endif ()
 
 set (MONGOC_ENABLE_SHM_COUNTERS 0)
 
-if (NOT ENABLE_SHM_COUNTERS MATCHES "ON|OFF|AUTO")
```
Comment (kevinAlbs) [inline src/libmongoc/CMakeLists.txt:375]: Can also remove:

```cmake
if (NOT ENABLE_SASL MATCHES "CYRUS|SSPI|AUTO|OFF")
   message (FATAL_ERROR
      "ENABLE_SASL option must be CYRUS, SSPI, AUTO, or OFF")
endif ()
```

**Review by kevinAlbs — COMMENTED** [review]
The added logs showing value changes seem very helpful.

Noted a slight concerned about raising the minimum cmake for users on older Debian versions that have older packaged cmake.

Noted a concern about changing the accepted values in options. If accepted option values are changed, I prefer the old value result in a configuration error. I want to prevent silently changing the build for existing build scripts.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,58 +1,174 @@
...
+         if(MSVC)
+            set(DEFAULT OFF)
+         else()
+            set(DEFAULT ON)
+         endif()
+      ]])
+
+# Toggle instrumentation:
+mongo_bool_setting(ENABLE_TRACING "Enable verbose debug output in the build driver"
+                   DEFAULT VALUE OFF AUDIT VALUE ON)
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply CMakeLists.txt:38]: Good point. This should not be on by any default.

The intention for `AUDIT` is to be a default CI option. Currently it isn't enabled, though.

File: build/cmake/MongoSettings.cmake [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+    # Developer-mode options:
+    if(DEFINED dflt_DEVEL AND MONGODB_DEVELOPER)
+        list(APPEND CMAKE_MESSAGE_CONTEXT "devel")
+        _mongo_compute_default(tmp "${dflt_DEVEL}")
+        message(DEBUG "Detected MONGODB_DEVELOPER: Default of ${setting_NAME} is “${tmp}”")
+        set("${outvar}" "${tmp}" PARENT_SCOPE)
+        return()
+    # Audit-mode options:
+    elseif(DEFINED dflt_AUDIT AND (MONGODB_DEVELOPER OR MONGODB_BUILD_AUDIT))
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply build/cmake/MongoSettings.cmake:213]: There was a difference when more options were updated, but for now, no.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,58 +1,174 @@
...
+list (APPEND CMAKE_MODULE_PATH
+      "${PROJECT_SOURCE_DIR}/build/cmake"
+      "${PROJECT_SOURCE_DIR}/build/cmake/make_dist"
+      )
+
+include (MongoSettings)
+
+# Subcomponents:
+mongo_bool_setting(ENABLE_MONGOC "Enable the build of the mongo-c-driver libraries")
+mongo_bool_setting(ENABLE_BSON "Enable the build of the C BSON handling libraries")
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply CMakeLists.txt:14]: I think it unlikely that many users are using `SYSTEM`, as that value is intended for package maintainers. For the sake of maintainers, I added the suggested message as a warning.

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply CMakeLists.txt]: Un-defaulted

**Review by rcsanchez97 — APPROVED** [review]
I have made a few comments for your consideration, but overall this looks good. Approving now so that if I am not able to review again in the next few days I am not holding up a merge.

LGTM

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,58 +1,195 @@
...
+mongo_bool_setting(ENABLE_MONGOC "Enable the build of the mongo-c-driver libraries")
+mongo_bool_setting(
+   USE_SYSTEM_LIBBSON [[Do not build libbson, and instead use an existing one via find_package()]]
+   DEFAULT VALUE OFF
+   ADVANCED
+)
+
+# Deprecated option "ENABLE_BSON" is gone, but old caches may rely on SYSTEM to
+# disable the libbson build. Allow ENABLE_BSON=SYSTEM only if USE_SYSTEM_LIBBSON
+# is also true, to allow both ENABLE_BSON=SYSTEM and USE_SYSTEM_LIBBSON=TRUE to
```
Comment (kevinAlbs) [inline CMakeLists.txt:22]: Allowing both options seems helpful. Consider adding a bullet to the unreleased section of the NEWS to warn users of the change:

> The `ENABLE_BSON` configuration option has been removed. Downstream build scripts using `ENABLE_BSON=SYSTEM` should be updated to use `USE_SYSTEM_LIBBSON=ON`. If the build script needs to support multiple versions of the C driver, add both `ENABLE_BSON=SYSTEM` and `USE_SYSTEM_LIBBSON=TRUE`.

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,58 +1,174 @@
...
-set (ENABLE_SNAPPY AUTO CACHE STRING "Enable snappy support. Set to ON/AUTO/OFF, default AUTO.")
-set (ENABLE_ZLIB AUTO CACHE STRING "Enable zlib support")
-set (ENABLE_ZSTD AUTO CACHE STRING "Enable zstd support. Set to ON/AUTO/OFF, default AUTO.")
-option (ENABLE_MAN_PAGES "Build MongoDB C Driver manual pages." OFF)
-option (ENABLE_HTML_DOCS "Build MongoDB C Driver HTML documentation." OFF)
-option (ENABLE_EXTRA_ALIGNMENT
-   "Turn on extra alignment of libbson types. Set to ON/OFF, default ON.\
- Required for the 1.0 ABI but better disabled."
-   ON
+cmake_minimum_required (VERSION 3.15)
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply CMakeLists.txt:1]: Thank you for the additional information. I am OK with bumping the minimum CMake.

I recommend updating the NEWS.

> It remains that the build-with-toolchain uses an old CMake version that conflicts with the newer CMake required. Addressing this is an open problem.

I expect the toolchain will need to be updated. The toolchain is built from this repo: https://github.com/10gen/mongo-c-toolchain as part of this Evergreen project: https://spruce.mongodb.com/commits/mongo-c-driver-toolchain

**Review by kevinAlbs — APPROVED** [review]
LGTM with open comments addressed.

File: build/cmake/MongoSettings.cmake [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+    file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/_eval.tmp.cmake" "${code}")
+    include("${CMAKE_CURRENT_BINARY_DIR}/_eval.tmp.cmake")
+    # Lift the variables into the caller's scope
+    foreach(__varname IN LISTS __eval_liftvars)
+        if(DEFINED "${__varname}")
+            message(TRACE "Eval variable result: ${__varname}=${${__varname}}")
+            set("${__varname}" "${${__varname}}" PARENT_SCOPE)
+        endif()
+    endforeach()
+endfunction()
```
Comment (eramongodb) [inline build/cmake/MongoSettings.cmake:284]:
```suggestion
endfunction()

```
Missing EOF newline.

File: build/cmake/MongoSettings.cmake [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        # Check that the default is actually a valid option:
+        if(DEFINED setting_OPTIONS
+            AND NOT DEFAULT IN_LIST setting_OPTIONS
+            AND NOT setting_ALLOW_OTHER_VALUES)
+            message(AUTHOR_WARNING "${setting_NAME}: Setting's default value is “${DEFAULT}”, which is not one of the provided setting options (${opts})")
+        endif()
+
+        # Reset "AUTO" values to the default
+        if(NOT "AUTO" IN_LIST setting_OPTIONS AND "$CACHE{${setting_NAME}}" STREQUAL "AUTO")
+            message(WARNING "Replacing ${setting_NAME}=“AUTO” with default value ${setting_NAME}=“${DEFAULT}”")
```
Comment (eramongodb) [inline build/cmake/MongoSettings.cmake:130]:
```suggestion
            message(WARNING "Replacing old ${setting_NAME}=“AUTO” with new default value ${setting_NAME}=“${DEFAULT}”")
```

**Review by eramongodb — COMMENTED** [review]
Great work! The `DEVEL` options pattern is very neat. For the most part, LGTM, but there are several suggestion and questions I would like to be addressed.

File: CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply CMakeLists.txt]: I went a surprise third route, and removed `ENABLE_BSON` entirely. Refer to fc15e0186565b57acf91d844207b3f82fe79ed85 for an explanation. @rcsanchez97 could you check this change in particular?

**Comment by rcsanchez97** [comment]
OK. So, I did some research and I think you should go a little further
and also remove the `USE_SYSTEM_LIBBSON` option you introduced.

The history here is that prior to 1.10.0 libbson and the C driver were
two separate projects. A user could clone libbson and build/install it
only. If you wanted the C driver, you had to first get/build/install
libbson and then get/build/install the C driver. When the two projects
were merged into a single source repository and a single CMake project,
we introduced the `ENABLE_BSON` option to allow the user to either build
libbson and the C driver together in one step, or separately in two
steps (by using the `SYSTEM` value to tell the C driver build in the
second step to go look for the libbson already present on the system and
the `BUNDLED` value to tell it to look for the libbson source in a
sub-directory in a one-step build).

The support for a two-step build, IIRC, was to allow users to keep
existing two-step build workflows intact. It also had to do with the
libbson repo being initially merged into the C driver repo as a
top-level project and then only later being made into a sub-project
(from the CMake perspective). However, I can't imagine that any of this
ia anything that we need to worry about that any longer (it has been
more than 5 years, after all). You can get a more complete picture of
the evolution by looking at CDRIVER-2505 and the commits associated with
it.

Keeping `ENABLE_MONGOC` around makes a lot of sense, since it is
possible that a user wants only libbson (either because libbson is the
only requirement for a given project, or it is the only required
dependency like for libmongocrypt). In effect, that makes
`ENABLE_MONGOC` necessary. However, there is no way to build the C
driver /without/ libbson and these days it seems like a vanishingly
small possibility that someone would want to build only the C driver
against an already existing libbson installation. The possibility is so
small, in my opinion, that is essesntially zero and not worth going to
the effort to keep around (even in the form of `USE_SYSTEM_LIBBSON`) if
you are already tweaking the option anyways. Better to just drop it
altogether.

In summary, there are basically two use cases that relate to the
potential need for `ENABLE_BSON` and/or `USE_SYSTEM_LIBBSON`.

1. A user wants to build only libbson without libmongoc
2. A user wants to build only libmongoc without libbson (i.e., using a
libbson previously built and installed on the system)

The first is accomplished with `ENABLE_MONGOC` and the second hasn't
been a thing for probably 5+ years and so keeping an option around for
it adds complexity and maintenance burden for no discernable benefit.

**Comment by vector-of-bool** [comment]
@rcsanchez97 How would this work for packagers that only want to build/install the libmongoc component against a previously-packaged libbson? e.g. Using `ENABLE_BSON=SYSTEM` is how it is currently done in vcpkg [Refer](https://github.com/microsoft/vcpkg/blob/1e334d770ac71a89d1c36d51019e60dd08b68f76/ports/mongo-c-driver/portfile.cmake).

I think there is more simplification to be done in this regard, but I think it may be for a later date. (Possible within the same scope of CDRIVER-4546).

File: CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply CMakeLists.txt:1]: Opened this: https://github.com/10gen/mongo-c-toolchain/pull/2

File: build/cmake/MongoSettings.cmake [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+    set("${setting_NAME}-PREV" "${${setting_NAME}}" CACHE INTERNAL "Prior value of ${setting_NAME}")
+
+    # Validation of options:
+    if(NOT setting_ALLOW_OTHER_VALUES AND (DEFINED setting_OPTIONS) AND (NOT ("${${setting_NAME}}" IN_LIST setting_OPTIONS)))
+        message(FATAL_ERROR "The value of “${setting_NAME}” must be one of [${opts}] (Got ${setting_NAME}=“${${setting_NAME}}”)")
+    endif()
+    if(setting_TYPE STREQUAL "BOOL"
+        AND NOT setting_ALLOW_OTHER_VALUES
+        AND NOT "${${setting_NAME}}" MATCHES "^(TRUE|FALSE|ON|OFF|YES|NO|0|1)$")
+        message(WARNING "The value of ${setting_NAME}=“${${setting_NAME}}” is not a regular boolean value")
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply build/cmake/MongoSettings.cmake:174]: I'm inclined to leave this. The `ENABLE_BSON` option is now handled separately.

File: build/cmake/MongoSettings.cmake [hunk 12]
```diff
@@ -0,0 +1,284 @@
+include_guard(DIRECTORY)
+
+#[[ Bool: Set to TRUE if the environment variable MONGODB_DEVELOPER is a true value ]]
+set(MONGODB_DEVELOPER FALSE)
+
+# Detect developer mode:
+set(_is_dev "$ENV{MONGODB_DEVELOPER}")
+if(_is_dev)
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply build/cmake/MongoSettings.cmake:8]: In most cases, yes, but this is the best way to avoid the wacky world of `if` parsing. The most predictable behavior is to activate if's `<variable>` parsing, but the only way to do that is with a bare identifier naming a variable. As such, I define a variable with the value of the env var, and then name that variable.

File: build/cmake/MongoSettings.cmake [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+    set_property(CACHE "${setting_NAME}" PROPERTY ADVANCED "${setting_ADVANCED}")
+    if(setting_OPTIONS)
+        set_property(CACHE "${setting_NAME}" PROPERTY STRINGS "${setting_OPTIONS}")
+    endif()
+
+    # Report what we set:
+    if(NOT DEFINED prev_val)
+        message(VERBOSE "Setting: ${setting_NAME} := “${${setting_NAME}}”")
+    elseif("${${setting_NAME}}" STREQUAL prev_val)
+        message(DEBUG "Setting: ${setting_NAME} := “${${setting_NAME}}” (Unchanged)")
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply build/cmake/MongoSettings.cmake:160]: Yes, but only because it's the least interesting of the three branches.

File: build/cmake/MongoSettings.cmake [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+    If specified, then `<code>` will be evaluated after the setting value is
+    defined. `<code>` may issue warnings and errors about the value of the
+    setting.
+
+ADVANCED
+    If specified, the cache variable will be marked as an advanced setting
+
+ALLOW_OTHER_VALUES
+    If *not* specified, this call will validate that the setting's value is
+    valid according to TYPE and OPTIONS.
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply build/cmake/MongoSettings.cmake:71]: The original idea was "here's some options you might want, but there could be others." I don't think that's a use case we have at the moment. Removed.

File: .evergreen/scripts/compile-unix.sh [hunk 15]
```diff
@@ -59,11 +59,12 @@ configure_flags_append_if_not_null() {
   fi
 }
 
+configure_flags_append --log-level=debug
+configure_flags_append --log-context
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply .evergreen/scripts/compile-unix.sh:63]: Replaced with a note.

File: README.rst [hunk 16]
```diff
@@ -47,6 +47,7 @@ information:
     - Examples: Windows 10 64-bit x86, Ubuntu 16.04 64-bit x86, macOS 11.0
 - C Compiler and version.
     - Examples: GCC 7.3.0, Visual Studio Community 2017, clang 3.9, XCode 9.3
+- Run CMake with ``--log-level=debug`` and ``--log-context`` for more verbose output.
```
[the remark this answers is not in the document]
Comment (vector-of-bool) [reply README.rst:50]: Since this is specifically for troubleshooting, I will keep it at "debug". (But not "trace", because that would generate a huge mess.)

File: .evergreen/config_generator/etc/utils.py [hunk 17]
```diff
@@ -23,11 +23,23 @@ class Task(EvgTask):
...
 
 # Automatically formats the provided script and invokes it in Bash.
-def bash_exec(script, **kwargs):
+def bash_exec(
+    script,
+    *,
+    include_expansions_in_env: Iterable[str] | None = None,
+    working_dir: str | None = None,
+    command_type: EvgCommandType | None = None,
+    env: Mapping[str, str] | None = None,
```
Comment (eramongodb) [inline .evergreen/config_generator/etc/utils.py:32]: Is this really necessary? The point of the `bash_exec` helper is to be a lightweight abstraction for the common `bash -c '<script'>` pattern; all other parameters are directly forwarded to `subprocess_exec` for it to parse and validate. I would like to avoid having to maintain compatibility/consistency of the parameter list with that of `subprocess_exec`.

File: CMakeLists.txt [hunk 18]
```diff
@@ -1,58 +1,201 @@
...
-   "Enables building of position independent code for static library components."
-   ON
+if(NOT WIN32)
+   mongo_bool_setting(ENABLE_PIC "Enable position-independent code when building static libraries (Not applicable on Windows)")
+elseif(DEFINED ENABLE_PIC)
+   # We are on Windows, and ENABLE_PIC is defined.
+   message(STATUS "ENABLE_PIC=“${ENABLE_PIC}” has no effect on Windows (All Windows code is position-independent)")
+else()
+   # We are on Windows, and ENABLE_PIC is not defined, which is normal
+endif()
```
Comment (eramongodb) [inline CMakeLists.txt:44]: Suggest using pattern consistent with new MongoSettings utilities, e.g.
```cmake
mongo_bool_setting(
   ENABLE_PIC "Enable position-independent code when building static libraries"
   DEFAULT EVAL [[
      if (WIN32)
         set(DEFAULT OFF)
      else ()
         set(DEFAULT ON)
      endif ()
   ]]
   VALIDATE CODE [[
      if (WIN32 AND DEFINED ENABLE_PIC)
         message(NOTICE "ENABLE_PIC="${ENABLE_PIC}" has no effect on Windows")
      endif ()
   ]]
)
```

File: CMakeLists.txt [hunk 19]
```diff
@@ -1,58 +1,201 @@
...
+   DEFAULT VALUE AUTO
+   VALIDATE CODE [[
+      if(ENABLE_SASL STREQUAL "SSPI" AND NOT WIN32)
+         message(WARNING "ENABLE_SASL=SSPI is only supported on Windows platforms")
+      endif()
+   ]]
+)
+
+mongo_bool_setting(
+   ENABLE_EXTRA_ALIGNMENT
```
Comment (eramongodb) [inline CMakeLists.txt:129]: Suggest relocating below alongside `ENABLE_AUTOMATIC_INIT_AND_CLEANUP` under "Deprecated options:".

**Review by eramongodb — APPROVED** [review]
I'd like to wait for https://github.com/10gen/mongo-c-toolchain/pull/2 to be merged so we can verify the `link-with-*` tasks pass with the proposed changes to avoid interfering with patch builds. Otherwise, LGTM.

Concerning `USE_SYSTEM_BSON`, if we can encourage package maintainers to simplify their handling of libbson/libmongoc to avoid dealing with the system BSON situation (e.g. by making the `libmongoc` package conflict with the `libbson` package due to unconditionally providing its own `libbson`), I think I would prefer going down that route. If we cannot (e.g. too disruptive, or legitimate use cases for system/external libbson dependencies), then this option should remain.

File: CMakeLists.txt (hunk 18 above)
Comment (vector-of-bool) [reply CMakeLists.txt:44]: The conditional is to not even define the setting on Windows, and detect if it has been explicitly set by the user, which means their explicit setting is being ignored. The suggested replacement would unconditionally issue the NOTICE on Windows since the option will always be defined when going through that code path.

File: .evergreen/config_generator/etc/utils.py (hunk 17 above)
Comment (vector-of-bool) [reply .evergreen/config_generator/etc/utils.py:32]: The purpose of explicitly enumerating the parameters we want is to early-detect breaking changes and enable tool assistance. If Shrub changes/removes parameters in a way that causes a mismatch, I want tooling to call out the incompatibility. Using a bare `**kwargs` is opaque to tools.

File: CMakeLists.txt (hunk 18 above)
Comment (vector-of-bool) [reply CMakeLists.txt:44]: It would be re-defined every time configure passes through `mongo_setting()`

**Comment by rcsanchez97** [comment]
> @rcsanchez97 How would this work for packagers that only want to build/install the libmongoc component against a previously-packaged libbson? e.g. Using `ENABLE_BSON=SYSTEM` is how it is currently done in vcpkg [Refer](https://github.com/microsoft/vcpkg/blob/1e334d770ac71a89d1c36d51019e60dd08b68f76/ports/mongo-c-driver/portfile.cmake).
> 

So, it looks like vcpkg handles things a bit differently. Their approach would still require the `USE_SYSTEM_LIBBSON` option. Their choice to build libbson in one place and then the C driver in another place (and then having that C driver build depend on the previously built libbson) is clearly a current instance of the two step build I described.

> I think there is more simplification to be done in this regard, but I think it may be for a later date. (Possible within the same scope of [CDRIVER-4546](https://jira.mongodb.org/browse/CDRIVER-4546)).

Cool. I'm looking forward to that.
